File: test/solutions/121.jl
function max_profit(prices::Vector{Int})::Int
    min_price = Inf
    max_profit = 0
    
    for price in prices
        if price < min_price
            min_price = price
        else
            profit = price - min_price
            if profit > max_profit
                max_profit = profit
            end
        end
    end
    
    return max_profit
end


@testset "Max Profit Tests" begin
    # Example 1
    prices1 = [7,1,5,3,6,4]
    expected_output1 = 5
    @test max_profit(prices1) == expected_output1
    
    # Example 2
    prices2 = [7,6,4,3,1]
    expected_output2 = 0
    @test max_profit(prices2) == expected_output2
    
    # Custom Test 1
    prices3 = [1,2,3,4,5,6,7,8,9,10]
    expected_output3 = 9
    @test max_profit(prices3) == expected_output3
    
    # Custom Test 2
    prices4 = [10,9,8,7,6,5,4,3,2,1]
    expected_output4 = 0
    @test max_profit(prices4) == expected_output4
end

File: test/rust/tests/121.rs
fn max_profit(prices: Vec<i32>) -> i32 {
    let mut max_profit = 0;
    let mut min_price = i32::MAX;

    for price in prices {
        min_price = std::cmp::min(min_price, price);
        max_profit = std::cmp::max(max_profit, price - min_price);
    }

    max_profit
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_example1() {
        let prices = vec![7, 1, 5, 3, 6, 4];
        assert_eq!(max_profit(prices), 5);
    }

    #[test]
    fn test_example2() {
        let prices = vec![7, 6, 4, 3, 1];
        assert_eq!(max_profit(prices), 0);
    }
}
